browser/ui/webui/new_tab_page/ms_edge_protocol_message_handler.cc

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) 2021 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// you can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "brave/browser/ui/webui/new_tab_page/ms_edge_protocol_message_handler.h"
+
+#include "base/bind.h"
+#include "base/command_line.h"
+#include "base/files/file_path.h"
+#include "base/logging.h"
+#include "base/notreached.h"
+#include "base/path_service.h"
+#include "base/task/thread_pool.h"
+#include "brave/browser/default_protocol_handler_utils_win.h"
+#include "chrome/installer/util/shell_util.h"
+
+namespace {
+
+constexpr wchar_t kMSEdgeProtocol[] = L"microsoft-edge";
+constexpr wchar_t kMSEdgeProtocolRegKey[] =
+    L"SOFTWARE\\Microsoft\\Windows\\Shell\\Associations\\UrlAssociations\\micro"
+    L"soft-edge";
+
+}  // namespace
+
+MSEdgeProtocolMessageHandler::MSEdgeProtocolMessageHandler()
+    : user_choice_key_(HKEY_CURRENT_USER, kMSEdgeProtocolRegKey, KEY_NOTIFY) {
+  StartWatching();
+}
+
+MSEdgeProtocolMessageHandler::~MSEdgeProtocolMessageHandler() = default;
+
+void MSEdgeProtocolMessageHandler::StartWatching() {
+  if (user_choice_key_.Valid()) {
+    user_choice_key_.StartWatching(
+        base::BindOnce(&MSEdgeProtocolMessageHandler::OnRegValChanged,
+                       base::Unretained(this)));
+  }
+}
+
+void MSEdgeProtocolMessageHandler::OnRegValChanged() {
+  CheckMSEdgeProtocolDefaultHandlerState();
+  StartWatching();
+}
+
+void MSEdgeProtocolMessageHandler::CheckMSEdgeProtocolDefaultHandlerState() {
+  base::ThreadPool::PostTaskAndReplyWithResult(
+      FROM_HERE, {base::MayBlock()},
```

Comment:
please add `base::TaskPriority::USER_VISIBLE` or the thread pool can take too long to execute it.